src/strategies/flex_window.py

```diff
@@ -291,37 +457,401 @@ def distribute_peak_shaving(self, timesteps):
+                total_power = (min_total_power + max_total_power) / 2
+                # reset soc
+                for i, b in enumerate(sim_batteries):
+                    b.soc = old_soc[i]
+
+                # calculate needed power to load battery
+                for ts_info in new_timesteps:
+                    cur_avail_power = total_power - ts_info["total_load"]
+                    for b in sim_batteries:
+                        if b.soc > 1 - self.EPS:
+                            # adready charged
+                            break
+
+                            cur_avail_power = (
+                                0
+                                if cur_avail_power < b.min_charging_power
+                                else cur_avail_power
+                            )
+                        if cur_avail_power > 0:
+                            b.load(
+                                self.interval, (cur_avail_power / len(sim_batteries))
+                            )["avg_power"]
+
+                safe = all(
+                    [b.soc >= (1 - self.EPS) for b in sim_batteries])
+
+                if safe:
+                    max_total_power = total_power
+                else:
+                    min_total_power = total_power
+            # actual charge
+            avail_power = total_power - gc.get_current_load()
+            for b_id, battery in self.world_state.batteries.items():
+                avail_power = (
+                    0 if avail_power < battery.min_charging_power else avail_power
+                )
+                charge = battery.load(self.interval, (avail_power / len(batteries)))["avg_power"]
+                gc.add_load(b_id, charge)
+                timesteps[0]["total_load"] += charge
+
+        else:
+            # discharge battery
+            no_window_timesteps = [
+                item for item in timesteps if item["window"] is False]
+            new_timesteps = []
+            for i, row in enumerate(no_window_timesteps):
+                if no_window_timesteps[i]["timestep_idx"] != i:
+                    break
+                new_timesteps.append(row)
+
+            min_total_power = -gc.max_power
+            max_total_power = gc.max_power
+
+            old_soc = [b.soc for b in sim_batteries]
+
+            while max_total_power - min_total_power > self.EPS:
+                total_power = (min_total_power + max_total_power) / 2
+
+                for i, b in enumerate(sim_batteries):
+                    b.soc = old_soc[i]
+
+                cur_time = self.current_time - self.interval
+                # calculate needed power to load battery
+                for ts_info in new_timesteps:
+
+                    cur_time += self.interval
+                    cur_needed_power = ts_info["total_load"] - total_power
+
+                    for b in sim_batteries:
+                        if b.soc <= self.DISCHARGE_LIMIT + self.EPS:
```

Comment:
(and other lines, especially 611): DISCHARGE_LIMIT is irrelevant for stationary batteries